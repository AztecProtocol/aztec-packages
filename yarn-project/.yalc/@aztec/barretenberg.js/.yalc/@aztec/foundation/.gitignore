dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs