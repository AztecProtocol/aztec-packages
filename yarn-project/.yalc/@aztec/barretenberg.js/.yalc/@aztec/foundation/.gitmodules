[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
