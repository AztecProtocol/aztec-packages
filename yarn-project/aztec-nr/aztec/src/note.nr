mod lifecycle;
mod note_getter;
mod note_getter_options;
mod note_hash;
mod note_header;
mod note_interface;
mod note_viewer_options;
mod utils;
