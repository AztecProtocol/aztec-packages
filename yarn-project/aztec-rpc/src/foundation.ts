import nodeCrypto from 'crypto';
import { Keccak } from 'sha3';

export const randomBytes = (len: number) => {
  return nodeCrypto.randomBytes(len) as Buffer;
};

export function keccak256(input: Buffer | string) {
  const inputBuf = typeof input === 'string' ? Buffer.from(input) : input;
  const hash = new Keccak(256);
  return hash.update(inputBuf).digest();
}
