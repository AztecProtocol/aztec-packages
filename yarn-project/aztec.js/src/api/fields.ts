export { Point, Fr, Fq, GrumpkinScalar } from '@aztec/foundation/fields';
