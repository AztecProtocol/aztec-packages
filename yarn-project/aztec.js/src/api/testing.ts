export { AnvilTestWatcher } from '../test/anvil_test_watcher.js';
export { RollupCheatCodes } from '../test/rollup_cheat_codes.js';
export { AztecCheatCodes } from '../test/aztec_cheat_codes.js';
export { EthCheatCodes } from '@aztec/ethereum/eth-cheatcodes';
export { CheatCodes } from '../test/cheat_codes.js';
