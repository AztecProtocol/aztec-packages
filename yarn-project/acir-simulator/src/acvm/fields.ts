import { AztecAddress, EthAddress, Fr } from '@aztec/foundation';

export type ACVMField = `0x${string}`;
export type ACVMWitness = Map<number, ACVMField>;

export const ZERO_ACVM_FIELD: ACVMField = `0x${Buffer.alloc(32).toString('hex')}`;
export const ONE_ACVM_FIELD: ACVMField = `0x${'00'.repeat(31)}01`;

function adaptBufferSize(originalBuf: Buffer) {
  const buffer = Buffer.alloc(32);
  if (originalBuf.length > buffer.length) {
    throw new Error('Buffer does not fit in 32 bytes');
  }
  originalBuf.copy(buffer, buffer.length - originalBuf.length);
  return buffer;
}

export function toACVMField(value: AztecAddress | EthAddress | Fr | Buffer | boolean | number): `0x${string}` {
  if (typeof value === 'boolean') {
    return value ? ONE_ACVM_FIELD : ZERO_ACVM_FIELD;
  }

  let buffer;

  if (Buffer.isBuffer(value)) {
    buffer = value;
  } else if (typeof value === 'number') {
    buffer = Buffer.alloc(32);
    buffer.writeUInt32BE(value, 28);
  } else {
    buffer = value.toBuffer();
  }

  return `0x${adaptBufferSize(buffer).toString('hex')}`;
}

export function fromACVMField(field: `0x${string}`): Fr {
  const buffer = Buffer.from(field.slice(2), 'hex');
  return Fr.fromBuffer(buffer);
}
