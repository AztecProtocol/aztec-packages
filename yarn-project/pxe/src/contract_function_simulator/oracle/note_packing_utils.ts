import { Fr } from '@aztec/foundation/fields';
import type { AztecAddress } from '@aztec/stdlib/aztec-address';
import type { Note } from '@aztec/stdlib/note';

// TS equivalent of the `NoteMetadata::from_raw_data` function in `aztec/src/note/note_metadata.nr`
function fromRawData(nonzeroNoteHashCounter: boolean, maybeNoteNonce: Fr): { stage: number; maybeNoteNonce: Fr } {
  if (nonzeroNoteHashCounter) {
    if (maybeNoteNonce.equals(Fr.ZERO)) {
      return { stage: 1, maybeNoteNonce }; // PENDING_SAME_PHASE
    } else {
      return { stage: 2, maybeNoteNonce }; // PENDING_PREVIOUS_PHASE
    }
  } else if (!maybeNoteNonce.equals(Fr.ZERO)) {
    return { stage: 3, maybeNoteNonce }; // SETTLED
  } else {
    throw new Error('Note has a zero note hash counter and no nonce - existence cannot be proven');
  }
}

/**
 * Packs a note in a format that is compatible with the default Packable implementation of the retrieved note.
 *
 * @dev Unlike the default Packable implementation, this function first constructs the note metadata from the inputs
 * and only after that it packs the retrieved note. Hence it doesn't map one to one with `RetrievedNote::pack()`.
 *
 * @param contractAddress - The address of the contract that owns the note
 * @param noteNonce - The nonce injected into the note hash preimage by kernels.
 * @param index - Optional index in the note hash tree. If undefined, indicates a transient note
 * @param note - The note content containing the actual note data
 * @returns The packed note as an array of field elements
 */
export function packAsRetrievedNote({
  contractAddress,
  noteNonce,
  index,
  note,
}: {
  contractAddress: AztecAddress;
  noteNonce: Fr;
  index?: bigint;
  note: Note;
}) {
  // If index is undefined, the note is transient which implies that the nonzero_note_hash_counter has to be true
  const noteIsTransient = index === undefined;
  const nonzeroNoteHashCounter = noteIsTransient ? true : false;

  // To pack the note as retrieved note we first need to reconstruct the note metadata.
  const noteMetadata = fromRawData(nonzeroNoteHashCounter, noteNonce);

  // Pack metadata first (stage and maybe_note_nonce), followed by the rest
  return [...note.items, contractAddress, new Fr(noteMetadata.stage), noteMetadata.maybeNoteNonce];
}
