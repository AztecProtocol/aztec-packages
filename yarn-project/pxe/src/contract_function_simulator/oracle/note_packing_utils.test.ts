import { Fr } from '@aztec/foundation/fields';
import { updateInlineTestData } from '@aztec/foundation/testing/files';
import { AztecAddress } from '@aztec/stdlib/aztec-address';
import { Note } from '@aztec/stdlib/note';

import { packAsRetrievedNote } from './note_packing_utils.js';

it('packs retrieved note', () => {
  const noteInfo = {
    contractAddress: AztecAddress.fromField(new Fr(1n)),
    noteNonce: new Fr(2n),
    index: undefined, // Transient note
    note: new Note([new Fr(3n), new Fr(4n)]),
  };

  const packed = packAsRetrievedNote(noteInfo);

  expect(packed).toMatchSnapshot();

  // Run with AZTEC_GENERATE_TEST_DATA=1 to update noir test data
  const fieldArrayStr = `[${packed.map(f => f.toString()).join(',')}]`;
  updateInlineTestData(
    'noir-projects/aztec-nr/aztec/src/note/retrieved_note.nr',
    'packed_retrieved_note_from_typescript',
    fieldArrayStr,
  );
});
