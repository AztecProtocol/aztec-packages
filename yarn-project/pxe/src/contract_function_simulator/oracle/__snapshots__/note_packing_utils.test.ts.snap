// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`packs retrieved note 1`] = `
[
  "0x0000000000000000000000000000000000000000000000000000000000000003",
  "0x0000000000000000000000000000000000000000000000000000000000000004",
  "0x0000000000000000000000000000000000000000000000000000000000000001",
  "0x0000000000000000000000000000000000000000000000000000000000000002",
  "0x0000000000000000000000000000000000000000000000000000000000000002",
]
`;
